### PR TITLE
SDKS-2346 Unbind device from AM Server

### PR DESCRIPTION
**Pull request by witrisna**
# JIRA Ticket

[SDKS-2346](https://bugster.forgerock.org/jira/browse/SDKS-2346)

# Description

Unbind device from AM Server

# Definition of Done Checklist:

- [ ] Acceptance criteria is met.
- [ ] All tasks listed in the user story have been completed.
- [ ] Coded to standards.
- [ ] Ensure backward compatibility.
- [ ] API reference docs is updated.
- [ ] Unit tests are written.
- [ ] Integration tests are written.
- [ ] e2e tests are written.
- [ ] Functional spec is written/updated.
- [ ] contains example code snippets.
- [ ] Change log updated.
- [ ] Documentation story is created and tracked.
- [ ] Tech debts and remaining tasks are tracked in separated ticket(s).

### CODE REVIEW

**Review by vahancouver — DISMISSED** [review]
Changes look good! 
The only thing I'm not sure about is the name `DeviceBindingRepository` (and `DefaultDeviceBindingRepository`). We already have similar name `DeviceRepository` for local storage. And the name `DeviceBindingRepository` doesn't really imply the server side of things. I have no suggestions at the moment but would like to discuss this.

File: forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteDeviceBindingRepository.kt [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+                    override fun onResponse(call: Call, response: Response) {
+                        continuation.resume(response)
+                    }
+                })
+                continuation.invokeOnCancellation {
+                    try {
+                        call.cancel()
+                    } catch (e: Exception) {
+                        //ignore
```
Comment (jeyanthanperiyasamy) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteDeviceBindingRepository.kt:68]: Can we log the exception ?

File: samples/app/src/main/java/com/example/app/userkeys/UserKeysRoute.kt [hunk 2]
```diff
@@ -66,9 +66,14 @@ fun UserKeysRoute(viewModel: UserKeysViewModel, openDrawer: () -> Unit) {
         )
     }
 
+    userKeysState.exception?.apply {
+        Alert(exception = this)
+        userKeysState.exception = null
+    }
+
     Column(modifier = Modifier.padding(16.dp)) {
         Topbar(heading = "User Keys", openDrawer)
```
Comment (jeyanthanperiyasamy) [inline samples/app/src/main/java/com/example/app/userkeys/UserKeysRoute.kt:75]: Heading can be like this 

User Keys - DeviceBind 

or like

DeviceBind - User Keys

File: forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteDeviceBindingRepository.kt [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+                    OkHttpClientProvider.getInstance().lookup(serverConfig)
+                val request = okhttp3.Request.Builder()
+                    .header(ACCEPT_API_VERSION, "resource=1.0")
+                    .url(getUrl(userKey))
+                    .delete().build()
+                val call = client.newCall(request)
+                call.enqueue(object : Callback {
+                    override fun onFailure(call: Call, e: IOException) {
+                        if (continuation.isCancelled) return
+                        continuation.resumeWithException(e)
```
Comment (jeyanthanperiyasamy) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteDeviceBindingRepository.kt:57]: we can log the exception here

**Comment by jeyanthanperiyasamy** [comment]
> Changes look good! The only thing I'm not sure about is the name `DeviceBindingRepository` (and `DefaultDeviceBindingRepository`). We already have similar name `DeviceRepository` for local storage. And the name `DeviceBindingRepository` doesn't really imply the server side of things. I have no suggestions at the moment but would like to discuss this.

> DeviceBindingRepository

i think i like the name ..Local and Remote DeviceBindingRepository

**Comment by jeyanthanperiyasamy** [comment]
minor comments...Looks good

**Review by jeyanthanperiyasamy — COMMENTED** [review]
minor comments...Looks good

File: forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/LocalDeviceRepository.kt [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+const val idKey = "id"
+const val userIdKey = "userId"
+const val kidKey = "kid"
+const val authTypeKey = "authType"
+const val userNameKey = "username"
+
+/**
+ * Helper class to save and retrieve EncryptedMessage
+ */
+internal class LocalDeviceRepository(context: Context,
```
Comment (jeyanthanperiyasamy) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/LocalDeviceRepository.kt:40]: Rename to LocalDeviceBindingRepository ...

File: forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteDeviceBindingRepository.kt (hunk 1 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteDeviceBindingRepository.kt]: Updated

File: samples/app/src/main/java/com/example/app/userkeys/UserKeysRoute.kt (hunk 2 above)
Comment (witrisna) [reply samples/app/src/main/java/com/example/app/userkeys/UserKeysRoute.kt]: Updated

File: forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/LocalDeviceRepository.kt (hunk 4 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/LocalDeviceRepository.kt]: Updated

File: forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteDeviceBindingRepository.kt (hunk 3 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteDeviceBindingRepository.kt:60]: Updated

**Comment by vahancouver** [comment]
> DeviceBindingRepository

@witrisna now that we have _LocalDeviceBindingRepository_ and _RemoteDeviceBindingRepository_, shouldn't the interface be also named to **DeviceBindingRepository** instead of _DeviceRepository_?

**Review by jeyanthanperiyasamy — DISMISSED** [review]

File: forgerock-auth/src/main/java/org/forgerock/android/auth/exception/ApiException.java [hunk 5]
```diff
@@ -61,4 +61,12 @@ public String getError() {
         return error;
     }
 
+    @Override
+    public String toString() {
```
Comment (jeyanthanperiyasamy) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/exception/ApiException.java:65]: my IDE warned to add this @NonNull, not a big deal...when i pulled this branch 

@Override
@NonNull
    public String toString()

File: forgerock-auth/src/main/java/org/forgerock/android/auth/exception/ApiException.java [hunk 6]
```diff
@@ -61,4 +61,12 @@ public String getError() {
         return error;
```
Comment (jeyanthanperiyasamy) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/exception/ApiException.java:61]: update copyright

File: forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteDeviceBindingRepository.kt [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+import java.io.IOException
+import java.net.HttpURLConnection
+import java.net.URL
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+
+private val TAG = RemoteDeviceBindingRepository::class.java.simpleName
+
+internal class RemoteDeviceBindingRepository(val serverConfig: ServerConfig = Config.getInstance().serverConfig) :
+    DeviceRepository {
```
Comment (jeyanthanperiyasamy) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteDeviceBindingRepository.kt:30]: interface name can be DeviceBindingRepository ?

File: forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteDeviceBindingRepository.kt [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    override suspend fun persist(userKey: UserKey) {
+        throw UnsupportedOperationException()
+    }
+
+    override suspend fun getAllKeys(): List<UserKey> {
+        throw UnsupportedOperationException()
+    }
+
+    override suspend fun delete(userKey: UserKey) {
+        val response = invoke(userKey);
```
Comment (jeyanthanperiyasamy) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteDeviceBindingRepository.kt:40]: Remove semi colon

**Review by jeyanthanperiyasamy — COMMENTED** [review]
minor comments

File: forgerock-auth/src/main/java/org/forgerock/android/auth/exception/ApiException.java (hunk 5 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/exception/ApiException.java]: Convert to Kotlin class

File: forgerock-auth/src/main/java/org/forgerock/android/auth/exception/ApiException.java (hunk 6 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/exception/ApiException.java]: Updated

File: forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteDeviceBindingRepository.kt (hunk 8 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteDeviceBindingRepository.kt]: Updated

File: forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteDeviceBindingRepository.kt (hunk 7 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/RemoteDeviceBindingRepository.kt]: Updated

**Review by vahancouver — DISMISSED** [review]
LGTM

**Review by spetrov — APPROVED** [review]
LGTM! 👍🏻

File: forgerock-auth/src/androidTest/java/org/forgerock/android/auth/callback/DeviceBindingListAndUnbind.java [hunk 9]
```diff
@@ -109,27 +111,140 @@ public void onException(Exception e) {
...
 
-        // Delete one of the keys and confirm it was deleted
-        FRListenerFuture<Void> future = new FRListenerFuture<>();
+        // Attempt to remove a key without being authenticated with it should fail
+        FRListenerFuture<Void> future = new FRListenerFuture<Void>();
         userKeys.delete(userKeys.loadAll().get(0), false, future);
-        future.get();
+        try {
+            future.get();
+            deletionSuccess[0]++;
```
Comment (witrisna) [inline forgerock-auth/src/androidTest/java/org/forgerock/android/auth/callback/DeviceBindingListAndUnbind.java:183]: You can just use fail() function, with fail(), it throws exception if the code reach that point.

Comment (spetrov) [reply forgerock-auth/src/androidTest/java/org/forgerock/android/auth/callback/DeviceBindingListAndUnbind.java]: Fixed.

**Review by spetrov — APPROVED** [review]

**Pull request merged**